chainer/functions/array/separate.py

```diff
@@ -1,5 +1,38 @@
-from chainer.functions.array import reshape
-from chainer.functions.array import split_axis
+from chainer import cuda
+from chainer import function_node
+from chainer.functions.array import stack
+from chainer.utils import type_check
+
+
+class Separate(function_node.FunctionNode):
+
+    """Function that separates a given array."""
+
+    def __init__(self, axis):
+        self.axis = axis
+
+    def check_type_forward(self, in_types):
+        type_check.expect(in_types.size() == 1)
+        x_type = in_types[0]
+        if self.axis >= 0:
+            type_check.expect(self.axis < x_type.ndim)
+        else:
+            type_check.expect(-self.axis <= x_type.ndim)
+
+    def forward(self, inputs):
+        x, = inputs
+        self._xp = cuda.get_array_module(x)
+        xs = self._xp.split(x, x.shape[self.axis], self.axis)
```

Comment:
How about following code?
```python
head  = (None,) * axis
ys = [x[head + (i,)] for i in six.moves.range(x.shape[self.axis])]
```

Reply:
Is this faster even in cupy?

Reply:
I tested the both with numpy but the performance is not different. I think `split` is simpler and better.

Reply:
OK, I understand.

Reply:
I got following result. My code is x10 faster when I use `numpy`.
```
In [38]: axis = 1
In [39]: xp = numpy
In [40]: x = xp.zeros((1000,1000))
In [41]: %timeit xs = xp.split(x, x.shape[axis], axis);ys = [xp.squeeze(y, axis) for y in xs]
100 loops, best of 3: 1.77 ms per loop

In [42]: %timeit head  = (None,) * axis; ys = [x[head + (i,)] for i in six.moves.range(x.shape[axis])]
1000 loops, best of 3: 253 µs per loop


In [46]: xp = cupy
In [47]: x = xp.zeros((1000,1000))

In [48]: %timeit xs = xp.split(x, x.shape[axis], axis);ys = [xp.squeeze(y, axis) for y in xs]
100 loops, best of 3: 2.68 ms per loop

In [50]: %timeit head  = (None,) * axis; ys = [x[head + (i,)] for i in six.moves.range(x.shape[axis])]
100 loops, best of 3: 3.05 ms per loop
```

Reply:
We need to call `chainer.utils.force_array`. It makes the result worse.

Reply:
Oh, Your right.